core/src/main/java/org/testcontainers/utility/DockerImageName.java

```diff
@@ -159,63 +176,86 @@ public String getRegistry() {
         return registry;
     }
 
+    /**
+     * @param newTag version tag for the copy to use
+     * @return an immutable copy of this {@link DockerImageName} with the new version tag
+     */
     public DockerImageName withTag(final String newTag) {
-        return new DockerImageName(rawName, registry, repo, new TagVersioning(newTag));
+        return new DockerImageName(rawName, registry, repo, new Versioning.TagVersioning(newTag), compatibleSubstituteFor);
     }
 
-    private interface Versioning {
-        boolean isValid();
-
-        String getSeparator();
+    /**
+     * Declare that this {@link DockerImageName} is a compatible substitute for another image - i.e. that this image
+     * behaves as the other does, and is compatible with Testcontainers' assumptions about the other image.
+     *
+     * @param otherImageName the image name of the other image
+     * @return an immutable copy of this {@link DockerImageName} with the compatibility declaration attached.
+     */
+    public DockerImageName asCompatibleSubstituteFor(String otherImageName) {
+        return asCompatibleSubstituteFor(DockerImageName.parse(otherImageName));
     }
 
-    @Data
-    private static class TagVersioning implements Versioning {
-        public static final String TAG_REGEX = "[\\w][\\w\\.\\-]{0,127}";
-        private final String tag;
-
-        TagVersioning(String tag) {
-            this.tag = tag;
-        }
-
-        @Override
-        public boolean isValid() {
-            return tag.matches(TAG_REGEX);
-        }
-
-        @Override
-        public String getSeparator() {
-            return ":";
-        }
-
-        @Override
-        public String toString() {
-            return tag;
-        }
+    /**
+     * Declare that this {@link DockerImageName} is a compatible substitute for another image - i.e. that this image
+     * behaves as the other does, and is compatible with Testcontainers' assumptions about the other image.
+     *
+     * @param otherImageName the image name of the other image
+     * @return an immutable copy of this {@link DockerImageName} with the compatibility declaration attached.
+     */
+    public DockerImageName asCompatibleSubstituteFor(DockerImageName otherImageName) {
+        return new DockerImageName(rawName, registry, repo, versioning, otherImageName);
```

Comment:
Consider adding `@With(AccessLevel.PRIVATE)` to `otherImageName`, so that we can do:
```suggestion
        return withOtherImageName(otherImageName);
```

Reply:
Nice, this works well (same for `withTag`).

N.B. I've upgraded the Lombok dependency so that we can use modern `@With` rather than deprecated `@Wither`